managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -848,6 +874,97 @@ private long consumedLedgerSize(long ledgerSize, long ledgerEntries, long consum
     }
 
     @Override
+    public synchronized void asyncTerminate(TerminateCallback callback, Object ctx) {
+        if (state == State.Fenced) {
+            callback.terminateFailed(new ManagedLedgerFencedException(), ctx);
+            return;
+        } else if (state == State.Terminated) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Ignoring request to terminate an already terminated managed ledger", name);
+            }
+            callback.terminateComplete(lastConfirmedEntry, ctx);
+            return;
+        }
+
+        log.info("[{}] Terminating managed ledger", name);
+        state = State.Terminated;
+
+        LedgerHandle lh = currentLedger;
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Closing current writing ledger {}", name, lh.getId());
+        }
+
+        mbean.startDataLedgerCloseOp();
+        lh.asyncClose((rc, lh1, ctx1) -> {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Close complete for ledger {}: rc = {}", name, lh.getId(), rc);
+            }
+            mbean.endDataLedgerCloseOp();
+            if (rc != BKException.Code.OK) {
+                callback.terminateFailed(new ManagedLedgerException(BKException.getMessage(rc)), ctx);
```

Comment:
I thought about that and my reasoning was: 
 1. If you want to terminate, anyway you don't want to publish anymore on the topic
 2. If you get error/timeout, you need to retry anyway to make sure it's really terminated

So in this case, if the caller does not retry the operation, the topic will stay terminated
until it gets unloaded, but I don't think that should represent a problem.